core/src/main/java/org/elasticsearch/common/settings/Setting.java

```diff
@@ -0,0 +1,394 @@
+     * that matches all settings with this prefix.
+     */
+    public boolean isGroupSetting() {
+        return false;
+    }
+
+    /**
+     * Returns the default values string representation for this setting.
+     * @param settings a settings object for settings that has a default value depending on another setting if available
+     */
+    public String getDefault(Settings settings) {
+        return defaultValue.apply(settings);
+    }
+
+    /**
+     * Returns <code>true</code> iff this setting is present in the given settings object. Otherwise <code>false</code>
+     */
+    public boolean exists(Settings settings) {
+        return settings.get(key) != null;
+    }
+
+    /**
+     * Returns the settings value. If the setting is not present in the given settings object the default value is returned
+     * instead.
+     */
+    public T get(Settings settings) {
+        String value = getRaw(settings);
+        try {
+            return parser.apply(value);
+        } catch (ElasticsearchParseException ex) {
+            throw ex;
+        } catch (Throwable t) {
+            throw new IllegalArgumentException("Failed to parse value [" + value + "] for setting [" + getKey() + "]", t);
+        }
+    }
+
+    /**
+     * Returns the raw (string) settings value. If the setting is not present in the given settings object the default value is returned
+     * instead. This is useful if the value can't be parsed due to an invalid value to access the actual value.
+     */
+    public String getRaw(Settings settings) {
+        return settings.get(key, defaultValue.apply(settings));
+    }
+
+    /**
+     * Returns <code>true</code> iff the given key matches the settings key or if this setting is a group setting if the
+     * given key is part of the settings group.
+     * @see #isGroupSetting()
+     */
+    public boolean match(String toTest) {
+        return Regex.simpleMatch(key, toTest);
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        builder.field("key", key);
+        builder.field("description", description);
+        builder.field("type", scope.name());
+        builder.field("dynamic", dynamic);
+        builder.field("default", defaultValue.apply(Settings.EMPTY));
+        builder.endObject();
+        return builder;
+    }
+
+    /**
+     * The settings scope - settings can either be cluster settings or per index settings.
+     */
+    public enum Scope {
+        Cluster,
+        Index;
+    }
+
+    SettingsService.SettingUpdater newUpdater(Consumer<T> consumer, ESLogger logger, Settings settings) {
+        return newUpdater(consumer, logger, settings, (s) -> true);
+    }
+
+    SettingsService.SettingUpdater newUpdater(Consumer<T> consumer, ESLogger logger, Settings settings, Predicate<T> accept) {
```

Comment:
no this won't work. The consumer is called once we apply the setting. The Predicate during the prepare phase. the consumer should only assign the value otherwise we loose transactionality